### DISCUSSION TITLE
ImGuiAl Logger integration: introduction and acknownledgement

### DISCUSSION THREAD

**Discussion by pthom** [pthom/imgui_bundle#72] [Announcements]
[ImGuiAl](https://github.com/leiradel/ImGuiAl) provides several ImGui widgets.

Amongst them, HelloImGui uses the logger that came with it.

* [C++ API](https://github.com/pthom/hello_imgui/blob/3f8d514ef55a6d6f7b8a16051a651317b94bb3da/src/hello_imgui/hello_imgui_logger.h)
* [Python API](https://github.com/pthom/imgui_bundle/blob/93531949f83750902251347cc413398caf1873fa/bindings/imgui_bundle/hello_imgui.pyi#L1570)

[Online emscripten demo](https://traineq.org/ImGuiBundle/emscripten/bin/demo_logger.html)

**Comment by leiradel** [comment]
Hey, ImGuiAl author here :)

I couldn't find it in the repository, I'd love to update my repository with the changes to use colors computed automatically. Was it removed from imgui_bundle?

**Comment by pthom** [comment]
Hey, thanks for contacting me! 

First of all, sorry for not having contacted you directly sooner, and for not having had time to send you a PR :-)

Actually, I am using ImGuiAl inside HelloImGui, which is a dependency of Dear ImGui Bundle. HelloImGui was developed two years before ImGui Bundle. I'm only using the logger (through imguial_term), and HelloImGui provides an API for it.

So, to answer your question, [this commit](https://github.com/pthom/hello_imgui/commit/425490c622fa7527bd069a2c68ea641154369f8d) in HelloImGui added automatic colors. Basically, it computes colors in the `HSV` color space:

* we have fixed hues for Debug_Blue, Info_Green, Warning_Yellow, Error_Red, and their hovered versions:

        constexpr float hBlue = 130.f / 255.f, hGreen = 94.f / 255.f,
                        hRed = 0.f, hYellow = 42.f / 255.f;
        constexpr float hBlueHovered = 117.f / 255.f, hGreenHovered = 56.f / 255.f,
                        hRedHovered = 228.f / 255.f, hYellowHovered = 36.f / 255.f;


* then we have makeColor which will create a HSV color depending on this hue and the luminance value of the background (`windowBgHsv_Value` below): the saturation is always one, but the luminance depends on the background.

```cpp
    ImU32 makeColor(LogColorClass logColorClass, LogColorState logColorState, float windowBgHsv_Value)
    {
        float wantedHue = logColorClassHue(logColorClass, logColorState);

        float adaptedValue = 1.f;
        if (windowBgHsv_Value < 0.8f)
            adaptedValue = 0.9f;
        else
            adaptedValue = 0.6f;

        float resultHsv[3] = { wantedHue, 1.f, adaptedValue };

        float resultRgb[3];
        ImGui::ColorConvertHSVtoRGB(resultHsv[0], resultHsv[1], resultHsv[2], resultRgb[0], resultRgb[1], resultRgb[2]);

        ImVec4 r { resultRgb[0], resultRgb[1], resultRgb[2], 1.f };
        return ImGui::GetColorU32(r);
    }
```

And here is the result on different backgrounds:

<img width="468" alt="image" src="https://user-images.githubusercontent.com/7694091/231758645-5a56fde2-7e77-4ee1-afe4-24b364ac9e40.png">

<img width="477" alt="image" src="https://user-images.githubusercontent.com/7694091/231758823-5c13a2c8-3ac9-4e67-a1a3-57a29d6de3e7.png">

Also, I make sure that the text color on the filter buttons is constant (and as you can see the text color on the red error button could still be improved)

I just created a PR to help you if you want to integrate this. I could not test that it compiles and works well on your repository since I did not recreate a working environment for compiling it. 
See https://github.com/leiradel/ImGuiAl/pull/2

**Comment by leiradel** [comment]
> First of all, sorry for not having contacted you directly sooner, and for not having had time to send you a PR :-)

No problem!

> I just created a PR to help you if you want to integrate this. I could not test that it compiles and works well on your repository since I did not recreate a working environment for compiling it.
> See https://github.com/leiradel/ImGuiAl/pull/2

Ah thanks a bunch! I'll check it out.
